compiler/rustc_parse/src/parser/item.rs

```diff
@@ -806,9 +843,29 @@ impl<'a> Parser<'a> {
         // Parse optional colon and param bounds.
         let bounds =
             if self.eat(&token::Colon) { self.parse_generic_bounds(None)? } else { Vec::new() };
-        generics.where_clause = self.parse_where_clause()?;
+
+        let before_where_clause = self.parse_where_clause()?;
```

Comment:
```suggestion
        generics.where_clause = self.parse_where_clause()?;
```
The predicates can then be extended in place.

Also it's better to skip the "after" `self.parse_where_clause()?` part if the current token is not `where` at that point, in that case `where` won't be added to the list of expected tokens (so UI tests like `src/test/ui/parser/bounds-lifetime-where.rs` won't change).